src/go/k8s/internal/controller/pvcunbinder/pvcunbinder.go

```diff
@@ -0,0 +1,313 @@
+	// allows disks to rebind if a Node happens to recover.
+	for _, pv := range pvs {
+		if err := r.recyclePersistentVolume(ctx, pv); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	missingPVCs := false
+	for _, pvc := range pvcByKey {
+		if pvc == nil {
+			missingPVCs = true
+			break
+		}
+	}
+
+	// 5. Delete the Pod to cause the StatefulSet controller to re-create both
+	// the PVCs and the Pod but only if there are missing PVCs.
+	if !missingPVCs {
+		return ctrl.Result{}, nil
+	}
+
+	logger.Info("deleting Pod to trigger PVC recreation", "name", pod.Name)
+	if err := r.Client.Delete(ctx, &pod, &client.DeleteOptions{
+		Preconditions: &metav1.Preconditions{
+			UID:             &pod.UID,
+			ResourceVersion: &pod.ResourceVersion,
+		},
+	}); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) ensureRetainPolicy(ctx context.Context, pv *corev1.PersistentVolume) error {
+	if pv.Spec.PersistentVolumeReclaimPolicy == corev1.PersistentVolumeReclaimRetain {
+		return nil
+	}
+
+	log.FromContext(ctx).Info("setting reclaim policy to retain", "name", pv.Name)
+
+	patch := client.StrategicMergeFrom(pv.DeepCopy(), &client.MergeFromWithOptimisticLock{})
+	pv.Spec.PersistentVolumeReclaimPolicy = corev1.PersistentVolumeReclaimRetain
+	if err := r.Client.Patch(ctx, pv, patch); err != nil {
+		return err
+	}
+	return nil
+}
+
+// recyclePersistentVolume "recycles" a released PV by clearing it's .ClaimRef
+// which makes it available for binding once again.
+// This strategy is only valid for volumes that utilize .HostPath.
+func (r *Reconciler) recyclePersistentVolume(ctx context.Context, pv *corev1.PersistentVolume) error {
+	if pv.Spec.HostPath == nil {
+		return fmt.Errorf("%T must specify a HostPath for recyling: %q", pv, pv.Name)
+	}
+
+	// Skip over unbound PVs.
+	if pv.Spec.ClaimRef == nil {
+		return nil
+	}
+
+	log.FromContext(ctx).Info("Releasing PersistentVolume", "name", pv.Name)
+
+	// NB: We explicitly don't use an optimistic lock here as the control plane
+	// will likely have updated this PV's Status to indicate that it's now
+	// Released.
+	patch := client.StrategicMergeFrom(pv.DeepCopy())
+	pv.Spec.ClaimRef = nil
+	if err := r.Client.Patch(ctx, pv, patch); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *Reconciler) shouldRemediate(ctx context.Context, pod *corev1.Pod) (bool, time.Duration) {
+	if r.Selector != nil && !r.Selector.Matches(labels.Set(pod.Labels)) {
+		log.FromContext(ctx).Info("selector not satisfied; skipping", "name", pod.Name, "labels", pod.Labels, "selector", r.Selector.String())
+		return false, 0
+	}
+
+	idx := slices.IndexFunc(pod.Status.Conditions, func(cond corev1.PodCondition) bool {
+		return cond.Type == corev1.PodScheduled && cond.Status == corev1.ConditionFalse && cond.Reason == "Unschedulable"
+	})
+
+	// Paranoid check, ensure that the Pod we've fetched still passes our predicate.
+	if idx == -1 || !pvcUnbinderPredicate(pod) {
```

Comment:
it does seem unnecessary indeed unless you refetch metadata for the pod from the k8s API.